client/src/unifyfs-sysio.c

```diff
@@ -886,22 +886,57 @@ off_t UNIFYFS_WRAP(lseek)(int fd, off_t offset, int whence)
 
         /* get current file position */
         off_t current_pos = filedesc->pos;
-
-        /* TODO: support SEEK_DATA and SEEK_HOLE? */
+        off_t logical_eof;
 
         /* compute final file position */
         switch (whence) {
         case SEEK_SET:
             /* seek to offset */
+            if (offset < 0) {
+                /* negative offset is invalid */
+                errno = EINVAL;
+                return (off_t)(-1);
+            }
             current_pos = offset;
             break;
         case SEEK_CUR:
             /* seek to current position + offset */
+            if (current_pos + offset < 0) {
+                /* offset is negative and will result in negative position */
+                errno = EINVAL;
+                return (off_t)(-1);
+            }
             current_pos += offset;
             break;
         case SEEK_END:
             /* seek to EOF + offset */
-            current_pos = unifyfs_fid_logical_size(fid);
+            logical_eof = unifyfs_fid_logical_size(fid);
+            if (logical_eof + offset < 0) {
+                /* offset is negative and will result in negative position */
+                errno = EINVAL;
+                return (off_t)(-1);
+            }
+            current_pos = logical_eof + offset;
+            break;
+        case SEEK_DATA:
```

Comment:
What does the team think of adding a ```#ifdef _GNU_SOURCE``` to conditionally compile these?  Worth doing that?

Reply:
I had that thought as well, and wouldn't be opposed. If so, does this effect how it is added in the Makefile (i.e., does it need to become a config option or anything)?